jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+			jfloatArray vectorArray = (jfloatArray)env->GetObjectArrayElement(vectors, i);
+			float* vector = env->GetFloatArrayElements(vectorArray, 0);
+			dim = env->GetArrayLength(vectorArray);
+			for(int j = 0; j < dim; ++j) {
+				dataset.push_back(vector[j]);
+			}
+			env->ReleaseFloatArrayElements(vectorArray, vector, 0);
+		}
+		has_exception_in_stack(env);
+
+		//---- indexPath
+		const char *indexString = env->GetStringUTFChars(indexPath, 0);
+        string indexPathString(indexString);
+        env->ReleaseStringUTFChars(indexPath, indexString);
+        has_exception_in_stack(env);
+
+		//---- algoParams
+		int paramsCount = env->GetArrayLength(algoParams);
+        for (int i=0; i<paramsCount; i++) {
+            jstring param = (jstring) (env->GetObjectArrayElement(algoParams, i));
+            const char *rawString = env->GetStringUTFChars(param, 0);
+            paramsList.push_back(rawString);
+
+            int M = 32;
+            if (sscanf(rawString, "M=%d", &M) == 1) {
+                indexDescription="HNSW"+std::to_string(M);
+            }
+            env->ReleaseStringUTFChars(param, rawString);
+
+        }
+		has_exception_in_stack(env);
+
+
+		//---- space
+		const char *spaceTypeCStr = env->GetStringUTFChars(spaceType, 0);
+        string spaceTypeString(spaceTypeCStr);
+        env->ReleaseStringUTFChars(spaceType, spaceTypeCStr);
+        has_exception_in_stack(env);
+		// space mapping faiss::MetricType
+		if(mapMetric.find(spaceTypeString) != mapMetric.end()) {
+			metric = mapMetric[spaceTypeString];
+		}
+
+		//---- Create IndexWriter from faiss index_factory
+		indexWriter.reset(faiss::index_factory(dim, indexDescription.data(), metric));
+
+		//Preparation And TODO Verify IndexWriter
+		//Some Param Can not Create from IndexFactory, Like HNSW efSearch and efCOnstruction
+		//----FOR HNSW 1st PARAM: M(HNSW32->M=32), efConstruction, efSearch
+		if(indexDescription.find("HNSW") != std::string::npos) {
+		    for(int i = 0; i < paramsCount; ++i) {
+		        const string& param = paramsList[i];
+		        int efConstruction = 40; //default
+		        int efSearch = 16;//default
+		        if(param.find("efConstruction") != std::string::npos &&
+		            sscanf(param.data(), "efConstruction=%d", &efConstruction) == 1) {
+                        faiss::IndexHNSW* ihp = reinterpret_cast<faiss::IndexHNSW*>(indexWriter.get());
+                        ihp->hnsw.efConstruction = efConstruction;
+		        } else if (param.find("efSearch") != std::string::npos &&
+                            sscanf(param.data(), "efSearch=%d", &efSearch) == 1){
+                        faiss::IndexHNSW* ihp = reinterpret_cast<faiss::IndexHNSW*>(indexWriter.get());
+                        ihp->hnsw.efSearch = efSearch;
+		        }
+		    }
+		}
+
+		//---- Do Index
+		//----- 1. Train
+        if(!indexWriter->is_trained) {
+			//TODO if we use like PQ, we have to train dataset
```

Comment:
Yeah I was thinking about this a little bit. It seems like it might make sense to fall back to a flat index if we do not have enough points to train. But we can think about this later.